src/Features/Core/Portable/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs

```diff
@@ -271,39 +275,72 @@ private bool IsDecomposableChainPart(SyntaxNode node)
         }
 
         /// <summary>
-        /// Recursively walks down <paramref name="node"/> decomposing it into the individual 
+        /// Walks down <paramref name="node"/> using a stack, decomposing it into the individual 
         /// tokens and nodes we want to look for chunks in. 
         /// </summary>
         private void Decompose(SyntaxNode node, ArrayBuilder<SyntaxNodeOrToken> pieces)
         {
-            // Ignore null nodes, they are never relevant when building up the sequence of
-            // pieces in this chained expression.
-            if (node is null)
+            var stack = SharedPools.Default<Stack<SyntaxNodeOrToken>>().AllocateAndClear();
+            try
             {
-                return;
-            }
+                SyntaxToken? token = null;
 
-            // We've hit some node that can't be decomposed further (like an argument list,
-            // or name node).  Just add directly to the pieces list.
-            if (!IsDecomposableChainPart(node))
-            {
-                pieces.Add(node);
-                return;
-            }
-
-            // For everything else that is a chain part, just decompose into its constituent 
-            // parts and add to the pieces array.
-            foreach (var child in node.ChildNodesAndTokens())
-            {
-                if (child.IsNode)
-                {
-                    Decompose(child.AsNode(), pieces);
-                }
-                else
+                // We'll start with the node that was passed in. If it was null, we will fall through
+                // since null nodes are never relevant when building up the sequence of pieces in chained
+                // expressions.
+                while (stack.Count > 0 || node != null)
```

Comment:
this feels weird.  why not unilaterally push teh node onto the stack.  then, we always just process the stack as the source of truth, and never need to have this sort of dual state check?